### PR TITLE
Rewrite C++ bits to Haskell

### LINKED ISSUE #184: Provide a static build?
Each of my travis builds now starts with
```
$ nix-env -iA cachix -f https://cachix.org/api/v1/install
unpacking 'https://cachix.org/api/v1/install'...
installing 'cachix-0.2.0'
these paths will be fetched (9.79 MiB download, 46.36 MiB unpacked):
  /nix/store/681354n3k44r8z90m35hm8945vsp95h1-glibc-2.27
  /nix/store/7mrq5v4nnqfkv1ka7a0kdw9mzvqs5w3c-gmp-6.1.2
  /nix/store/bnbq85lsbds5asi8vzjqhvrgq3l9ila8-cachix-0.2.0
  /nix/store/hlnxw4k6931bachvg5sv0cyaissimswb-gcc-7.4.0-lib
  /nix/store/iiymx8j7nlar3gc23lfkcscvr61fng8s-zlib-1.2.11
  /nix/store/k0y9nrym785l61pjw0lp0qk90v2nbpqs-ncurses-6.1-20190112
  /nix/store/kbbiky87bsdndh63m7919sxkf3gsh5zq-libffi-3.2.1
  /nix/store/mibh7947iwgf5yiqc7njn0igfzlxzdf1-xz-5.2.4
  /nix/store/y1ck4skkvmcnar0y0i1kkmvv0lasvy5b-cachix-0.2.0-data
copying path '/nix/store/y1ck4skkvmcnar0y0i1kkmvv0lasvy5b-cachix-0.2.0-data' from 'https://cache.nixos.org'...
copying path '/nix/store/681354n3k44r8z90m35hm8945vsp95h1-glibc-2.27' from 'https://cache.nixos.org'...
copying path '/nix/store/hlnxw4k6931bachvg5sv0cyaissimswb-gcc-7.4.0-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/kbbiky87bsdndh63m7919sxkf3gsh5zq-libffi-3.2.1' from 'https://cache.nixos.org'...
copying path '/nix/store/7mrq5v4nnqfkv1ka7a0kdw9mzvqs5w3c-gmp-6.1.2' from 'https://cache.nixos.org'...
copying path '/nix/store/k0y9nrym785l61pjw0lp0qk90v2nbpqs-ncurses-6.1-20190112' from 'https://cache.nixos.org'...
copying path '/nix/store/mibh7947iwgf5yiqc7njn0igfzlxzdf1-xz-5.2.4' from 'https://cache.nixos.org'...
copying path '/nix/store/iiymx8j7nlar3gc23lfkcscvr61fng8s-zlib-1.2.11' from 'https://cache.nixos.org'...
copying path '/nix/store/bnbq85lsbds5asi8vzjqhvrgq3l9ila8-cachix-0.2.0' from 'https://cache.nixos.org'...
building '/nix/store/rrdzli02y27g1y4ld8ckgsk3f66yrvay-user-environment.drv'...
created 19 symlinks in user environment
The command "nix-env -iA cachix -f https://cachix.org/api/v1/install" exited with 0.
```
ok, it is only 15 seconds, but still: Have you considered building `cachix` statically, maybe with `musl`, i.e. `crossSystem = localLib.systems.examples.musl64`, so that this step has to download only one self-contained and hopefully smaller store path?

### PR DESCRIPTION
**Pull request by domenkozar**
Fixes #184 #374

### CODE REVIEW

**Comment by roberth** [comment]
Oh wow I was going to send some warning fixes because I'd started to include cachix in my local build for tooling reasons. Seems that I may not have to bother with that? Anyway, that doesn't really matter, but...

Heaving read through the diff I'm concerned that doing it this way could cause trouble for you and your users. For instance, if you don't get the sqlite flags from the nix config correctly, you'll trigger #475 again. Also, you'll have to deal with changes to the db schema yourself. Probably other unknowns later.
Meanwhile, we're making progress on language bindings in nix upstream. A contributor is committing to C and Python bindings, so we'll have a good foundation for hsc2hs based bindings: no more template haskell and no more C++ interactions either.
Pick your poison :)

**Comment by domenkozar** [comment]
Seems like we need to pick the right vfs: https://github.com/NixOS/nix/blob/master/src/libstore/sqlite.cc#L58

I'm not happy about this direction, but we had way too much trouble with C++ and this part of the codebase seems quite stable for now. In the future we can reverse this direction if the bindings become more sane.

File: cachix/src/Cachix/Client/Store.hs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    go [SQLite.SQLText path_] = return path_
+    go a = throwIO $ FatalError $ "invalid reference type " <> show a
+
+    getDeriver :: SQLite.SQLData -> Maybe Text
+    getDeriver (SQLite.SQLText deriver) = Just deriver
+    getDeriver _ = Nothing
+
+computeClosure :: Store -> [StorePath] -> IO [StorePath]
+computeClosure store initialPaths = do
+  allPaths <- processGraph (getPath <$> initialPaths) (fmap (concatMap references . rights . List.singleton) . queryPathInfo store)
```
Comment (sandydoo) [inline cachix/src/Cachix/Client/Store.hs:102]: Do we care about any of the errors returned by `queryPathInfo` here? If not, should we document the reason?

What if we avoid going through a list here? The `.` could also be a `<$>` to be read as the common double fmap pattern.
```suggestion
  let fetchReferences = fmap (either (const []) references) . queryPathInfo store
  allPaths <- processGraph (getPath <$> initialPaths) fetchReferences
```

Or maybe even something really obvious:
```suggestion
  allPaths <- 
    processGraph (getPath <$> initialPaths) $
      queryPathInfo store >=> \case
        Left _ -> pure []
        Right pathInfo -> pure $ references pathInfo
```

File: cachix/src/Cachix/Client/Store.hs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import qualified Data.Text as T
+import Database.SQLite3 (SQLData)
+import qualified Database.SQLite3 as SQLite
+import Protolude hiding (toS)
+import Protolude.Conv
+import System.Directory (canonicalizePath)
+import qualified System.Nix.Base32
+import System.Process (readProcessWithExitCode)
+
+data Store = Store Text SQLite.Database
```
Comment (sandydoo) [inline cachix/src/Cachix/Client/Store.hs:18]: We should document that the `Text` is a store prefix. Maybe a type alias?

**Review by sandydoo — CHANGES_REQUESTED** [review]
👌

File: cachix/src/Cachix/Client/Store.hs [hunk 3]
```diff
@@ -99,23 +102,23 @@ queryPathInfo store path = do
...
 
 getPath :: StorePath -> Text
 getPath (StorePath storePath) = storePath
 
-getStorePathBaseName :: StorePath -> Text
-getStorePathBaseName (StorePath storePath) =
-  T.drop 11 storePath
+getStorePathBaseName :: Store -> StorePath -> Text
+getStorePathBaseName (Store storePrefix _) (StorePath storePath) =
+  T.dropPrefix storePrefix storePath
```
Comment (sandydoo) [inline cachix/src/Cachix/Client/Store.hs:123]: The default prefix is `/nix`, so this isn't dropping the `/store/` bit. We might also have to sanitize the prefix to handle any trailing slashes.

**Pull request merged**